log/file_test.go

```diff
@@ -0,0 +1,161 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type nopCloser struct {
+	io.Writer
+}
+
+func (nc *nopCloser) Close() error {
+	return nil
+}
+
+func TestFileHookFromConfigLine(t *testing.T) {
+	t.Parallel()
+
+	tests := [...]struct {
+		line       string
+		err        bool
+		errMessage string
+		res        fileHook
+	}{
+		{
+			line: "file",
+			err:  true,
+			res: fileHook{
+				levels: logrus.AllLevels,
+			},
+		},
+		{
+			line: "file=/tmp/k6.log,level=info",
+			err:  false,
+			res: fileHook{
+				path:   "/tmp/k6.log",
+				levels: logrus.AllLevels[:5],
+			},
+		},
+		{
+			line: "file=./",
+			err:  true,
+		},
+		{
+			line: "file=/a/c/",
+			err:  true,
+		},
+		{
+			line:       "file=,level=info",
+			err:        true,
+			errMessage: "filepath must not be empty",
+		},
+		{
+			line: "file=/tmp/k6.log,level=tea",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,unknown",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,level=",
+			err:  true,
+		},
+		{
+			line: "file=/tmp/k6.log,level=,",
+			err:  true,
+		},
+		{
+			line:       "file=/tmp/k6.log,unknown=something",
+			err:        true,
+			errMessage: "unknown logfile config key unknown",
+		},
+		{
+			line:       "unknown=something",
+			err:        true,
+			errMessage: "logfile configuration should be in the form `file=path-to-local-file` but is `unknown=something`",
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.line, func(t *testing.T) {
+			t.Parallel()
+
+			res, err := FileHookFromConfigLine(context.Background(), logrus.New(), test.line, make(chan struct{}))
+
+			if test.err {
+				require.Error(t, err)
+
+				if test.errMessage != "" {
+					require.Equal(t, test.errMessage, err.Error())
+				}
+
+				return
+			}
+
+			require.NoError(t, err)
+			assert.NotNil(t, res.(*fileHook).w)
+		})
+	}
+}
+
+func TestFileHookFire(t *testing.T) {
+	t.Parallel()
+
+	var buffer bytes.Buffer
+	nc := &nopCloser{
+		Writer: &buffer,
+	}
+
+	hook := &fileHook{
+		dataCh:          make(chan []byte),
+		loggerStoppedCh: make(chan struct{}),
+		w:               nc,
+		bw:              bufio.NewWriter(nc),
+		levels:          logrus.AllLevels,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	hook.dataCh = hook.loop(ctx)
+
+	logger := logrus.New()
+	logger.AddHook(hook)
+	logger.SetOutput(io.Discard)
+
+	logger.Info("example log line")
+
+	cancel()
+	<-hook.dataCh
+
+	assert.Contains(t, buffer.String(), "example log line")
```

Comment:
This test now fails. This (IMO) is because we now have a buffered channel between logger.Info -> ... -> fileHook.Fire and the actual place it's logged (the loop). 

With a buffered channel (as long as there is space) the send (in `Fire`) can just put it in the channel and continue. With an unbuffered channel (as before) the send needs to wait for the actual receive on the channel to happen, before it can continue. 

The latter will always manage to enter the `select` and write the message, the first won't. I don't think we care about losing the last few messages in some cases, as in the real case that *should* not be possible. But we still need this test to pass, so please add a small sleep and do `-race -count 10` a few times locally to be sure it's enough. I would wager anything above 1-10ms should be sufficient

Reply:
With the timeout, it works as expected. Thanks!
There are a few linter errors:
```
log/file.go:21:1    stylecheck  ST1000: at least one file in a package should have a package comment
log/levels.go:21:1  stylecheck  ST1000: at least one file in a package should have a package comment
```
Should I worry about them?

Reply:
@alyakimenko Yes, please. The linter must be fixed for getting the PR merged.

Reply:
@codebien
`Package log implements various logrus hooks.`
Do you find the description suitable for the package?

Reply:
@alyakimenko it's what it does so it's fine for me, but other reviewers could have different opinions